### ISSUE TITLE
Writing and reading WKB loses M dimension

### ISSUE THREAD

**Issue by kaiwinter** [locationtech/jts#733]
When using `WKBWriter` to write a Geometry which has XYM coordinates and afterwards using `WKBReader` for reading the result. The Geometry has XYZ coordinates instead of XYM coordinates.

Here is small example:
```
    String wkt = "MULTILINESTRING M((1 1 1, 2 2 2))";
    WKTReader wktReader = new WKTReader();
    Geometry geometryBefore = wktReader.read(wkt);

    WKBWriter wkbWriter = new WKBWriter(3);
    byte[] write = wkbWriter.write(geometryBefore);

    WKBReader wkbReader = new WKBReader();
    Geometry geometryAfter = wkbReader.read(write);

    System.out.println(Arrays.asList(geometryBefore.getCoordinates()));
    System.out.println(Arrays.asList(geometryAfter.getCoordinates()));
```

When running this you get the following output:
```
[(1.0, 1.0 m=1.0), (2.0, 2.0 m=2.0)]
[(1.0, 1.0, 1.0), (2.0, 2.0, 2.0)]

```
The `geometryBefore` has XYM coordinates. The `geometryAfter` has XYZ coordinates. 
In [`WKBReader.readGeometry`](https://github.com/locationtech/jts/blob/master/modules/core/src/main/java/org/locationtech/jts/io/WKBReader.java#L203) are two variables evaluated [`hasZ`](https://github.com/locationtech/jts/blob/master/modules/core/src/main/java/org/locationtech/jts/io/WKBReader.java#L241) and [`hasM`](https://github.com/locationtech/jts/blob/master/modules/core/src/main/java/org/locationtech/jts/io/WKBReader.java#L244). In the above example `hasZ` is true and `hasM` is false. I think this is because [`WKBWriter.writeGeometryType`](https://github.com/locationtech/jts/blob/master/modules/core/src/main/java/org/locationtech/jts/io/WKBWriter.java#L405) always writes `0x80000000` regardless of the coordinate type. Debugging `WKBReader.readGeometry` makes me think `0x40000000` should be used for M coordinates in `WKBWriter.writeGeometryType`.

What do you think? Can this be improved or did I just misunderstood how this should work.

**Comment by jnh5y** [comment]
I think you've identified a bug here.  Any interest in working on it?

The Contributing guide has some info about that here: https://github.com/locationtech/jts/blob/master/CONTRIBUTING.md

**Comment by kaiwinter** [comment]
Yes sure. Currently I'm using a workaround with a `GeometryFactory` but I would love to get rid of it as it most probably impacts performance.

When thinking of a fix my first thought was to do an `instanceof` check of the result of `g.getCoordinate()` in [`writeGeometry`](https://github.com/locationtech/jts/blob/jts-1.18.1/modules/core/src/main/java/org/locationtech/jts/io/WKBWriter.java#L408). But that would be pretty ugly. Is there a better way to get the subtype of the `Coordinate` (whether it is `CoordinateXYM`)?

When looking at the `WKTWriter` there is the method [`setOutputOrdinates()` ](https://github.com/locationtech/jts/blob/jts-1.18.1/modules/core/src/main/java/org/locationtech/jts/io/WKTWriter.java#L331). I was using this method to correctly write XYM coordinates as text, so I think this pattern may be applied to the `WTBWriter` as well. Do you have other ideas?

**Comment by TimPigden** [comment]
what's the situation here? This code is clearly wrong:

**Comment by TimPigden** [comment]
ok, i can get round it with
val wktWriter = new WKTWriter(3)
    wktWriter.setOutputOrdinates(Ordinate.createXYM()) // Include M dimension in

but clearly that's not an optimal solution - there's no reason for the M to require the dimensions to be set in the first place, surely. You're just making hurdles unnecessarily - if i want to createXYM then obviously i want 3 dimensions at that point
(and why does it really need it at all? It took digging around in the source to find all this. asText should just recognise I've got M in my dimensions and coordinates. Writing ought to be the default position I'd have thought - who wants an m and not wants to see it? Suppressing it should be the active decision)

**Comment by jodygarnett** [comment]
I have not seen a PR but would like to see this addressed. There is some utility methods to determine hasZ or hasM.

**Comment by kaiwinter** [comment]
There was a PR (#734) which has already been merged. The change should have been released with 1.20.0. This issue could be closed as resolved, IMHO.

**Comment by kaiwinter** [comment]
This library has its rough edges, without question. But keep in mind this is open source software and we can all contribute to make it continuously better. 

I'm not the right person to answer your question, I'm afraid. Maybe you should start a discussion or, if you have a good solution to your specific problem, describe it in an issue or a PR.
